### ISSUE TITLE
Openstack/os_router fails with "interfaces" option

### ISSUE THREAD

**Issue by parberge** [ansible/ansible#18915]
_From @peerster on October 7, 2016 12:23_

##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

os_router.py
##### ANSIBLE VERSION

```
$ ansible --version
ansible 2.1.2.0
```
##### OS / ENVIRONMENT

CentOS7 with Openstack Mitaka release
##### SUMMARY

Getting a fatal error when using the paramter `interfaces`
##### STEPS TO REPRODUCE

1. First setup a subnet for the project TEST called `Private Subnet`.
2. Define following vars for authentication

```
    test_auth_dict:
      password: my-password
      auth_url: http://my-url/v2.0
      username: test
      project_name: TEST
```

Then run following task:

```
  - name: "Setup router for test project"
    os_router:
      auth: '{{ test_auth_dict }}'
      name: Router 1
      state: present
      interfaces: Private Subnet
```

##### EXPECTED RESULTS

A router created with an interface connected to the subnet specified
##### ACTUAL RESULTS

```
TASK [Setup router for DCO project] ********************************************
<10.207.80.8> ESTABLISH SSH CONNECTION FOR USER: root
<10.207.80.8> SSH: EXEC sshpass -d12 ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o User=root -o ConnectTimeout=10 -o ControlPath=/home/my-user/.ansible/cp/ansible-ssh-%h-%p-%r 10.207.80.8 '/bin/sh -c '"'"'LANG=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8  LC_ALL=en_US.UTF-8 /usr/bin/python && sleep 0'"'"''
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_xX7nkA/ansible_module_os_router.py", line 367, in <module>
    main()
  File "/tmp/ansible_xX7nkA/ansible_module_os_router.py", line 323, in main
    if _needs_update(cloud, module, router, net, internal_ids):
  File "/tmp/ansible_xX7nkA/ansible_module_os_router.py", line 202, in _needs_update
    for port in cloud.list_router_interfaces(router, 'internal'):
  File "/usr/lib/python2.7/site-packages/shade/openstackcloud.py", line 2507, in list_router_interfaces
    'external_fixed_ips' in router['external_gateway_info']):
TypeError: argument of type 'NoneType' is not iterable

fatal: [my-host]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_name": "os_router"}, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_xX7nkA/ansible_module_os_router.py\", line 367, in <module>\n    main()\n  File \"/tmp/ansible_xX7nkA/ansible_module_os_router.py\", line 323, in main\n    if _needs_update(cloud, module, router, net, internal_ids):\n  File \"/tmp/ansible_xX7nkA/ansible_module_os_router.py\", line 202, in _needs_update\n    for port in cloud.list_router_interfaces(router, 'internal'):\n  File \"/usr/lib/python2.7/site-packages/shade/openstackcloud.py\", line 2507, in list_router_interfaces\n    'external_fixed_ips' in router['external_gateway_info']):\nTypeError: argument of type 'NoneType' is not iterable\n", "module_stdout": "", "msg": "MODULE FAILURE"}

```

Have changed/obfuscated some to remove unecessary information not related to the problem

_Copied from original issue: ansible/ansible-modules-core#5188_

**Comment by parberge** [comment]
_From @ansibot on October 7, 2016 13:11_



@emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by parberge** [comment]
_From @jwoltman on October 13, 2016 8:37_

Same issue here. First run is successful but second run fails.     'external_fixed_ips' in router['external_gateway_info']): TypeError: argument of type 'NoneType' is not iterable

ansible 2.1.2.0

**Comment by parberge** [comment]
_From @rcarrillocruz on October 13, 2016 13:11_

Happy to take a look at this.

**Comment by parberge** [comment]
_From @rcarrillocruz on October 14, 2016 10:10_

I'm unable to reproduce:

```
openstack network create test-net
openstack subnet create --network test-net --subnet-range 192.168.1.0/24 test-subnet
(venv) vagrant@vagrant-ubuntu-trusty-64:~$ cat test_os_router_interfaces.yml 
- hosts: localhost
  tasks:
    - name: "Setup router for test project"
      os_router:
        cloud: devstack-admin
        name: Router 1
        state: present
        interfaces: test-subnet
(venv) vagrant@vagrant-ubuntu-trusty-64:~$ ansible-playbook test_os_router_interfaces.yml 
 [WARNING]: Host file not found: /etc/ansible/hosts

 [WARNING]: provided hosts list is empty, only localhost is available


PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [Setup router for test project] *******************************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0   

```

**Comment by parberge** [comment]
@rcarrillocruz I can see you have cloud: devstack-admin.
I'm not really familiar with `cloud` parameter. 

In my example I'm authenticating as a user (non-admin) which is part of a project (not admin project).
Maybe there is something there which causes the problem?

**Comment by parberge** [comment]
@rcarrillocruz I did some more testing. It seems to be related to permissions.
If I authenticate as admin instead of the project user the task is successful.
However, the router will be in wrong project (admin), and that's not what I want.
I see that the parameter `project` was added to ansible 2.2, but I can't use it since I use 2.1.

So, is there any way to solve it in 2.1? If not is it possible to backport the option to 2.1? If no on both questions I guess this issue can be closed with 'Won't fix' or something

**Comment by parberge** [comment]
_From @ansibot on November 3, 2016 8:32_

@emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by parberge** [comment]
_From @ansibot on November 18, 2016 9:31_

@emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz, @thingee, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by parberge** [comment]
_From @ansibot on December 7, 2016 19:16_

This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by ansibot** [comment]
cc @shrews, @thingee
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by Thingee** [comment]
Hi @peerster! The cloud parameter is just a way to specify a profile to use for authentication. You can read more about in the [documentation](http://docs.openstack.org/developer/os-client-config/).

The permissions of creating a router with an external gateway can vary depending on what the policy is set to. You can see the default permission is for admin only in the [Neutron repository](http://git.openstack.org/cgit/openstack/neutron/tree/etc/policy.json#n51). I would say at the very least it shouldn't stack trace like it is and give a more useful error message. Let me take a closer look at things.

**Comment by parberge** [comment]
@Thingee Thanks (a bit late reponse, didn't get notified).

As it works using the `project` parameter in ansible 2.2.x it's no longer an issue for me.
But I agree that it should give a more useful error message.
